### PR TITLE
fix: Adds tests for safe derivation levels and explicitly checks 8*ZL against 2^255  ENG-440

### PR DESCRIPTION
**Pull request by HashMapsData2Value**
- Adds a BigIntException with Overflow case, representing when the highest bit (Bit 255) gets set.
- Adds a test that confirms that there are 8 safe levels for Peikert and 67108864 for Khovratovich. The latter takes some time to execute and so is not part of the normal tests. Test works by mocking deriveChildNodePrivate, specifically changing the deriveNonHardened call to always return 32 bytes of 0xFF.
- Also adds the safe levels test for known seed, which checks that we get overflow when we go from 24->25 levels (with the first 5 conforming to the normal Bip44 and 0s).
- Since we now explicitly trigger an exception, we need to add throws and tries.

### CODE REVIEW

File: Sources/bip32-ed25519-swift/bip32-ed25519-swift.swift [hunk 1]
```diff
@@ -227,14 +231,16 @@ public class Bip32Ed25519 {
...
         let left = BigUInt(Data(kl.reversed())) + BigUInt(Data(trunc256MinusGBits(zl: zl, g: g).reversed())) * BigUInt(8)
+        guard left < (BigUInt(1) << 255) else {
+            throw BigIntException.overflow
+        }
         let right = BigUInt(Data(kr.reversed())) + BigUInt(Data(zr.reversed()))
 
         // Reverse byte order back after calculations
         var leftData = Data(left.serialize().reversed())
-        var rightData = Data(right.serialize().reversed())
+        var rightData = Data(right.serialize().reversed()).prefix(ED25519_SCALAR_SIZE)
```
Comment (ehanoc) [inline Sources/bip32-ed25519-swift/bip32-ed25519-swift.swift:243]: Do we to prefix with 32bytes? What does it do?

Comment (kylebeee) [reply Sources/bip32-ed25519-swift/bip32-ed25519-swift.swift:243]: prefix just takes a slice of the specified length

File: Sources/bip32-ed25519-swift/bip32-ed25519-swift.swift [hunk 2]
```diff
@@ -227,14 +231,16 @@ public class Bip32Ed25519 {
         let zr = z.subdata(in: ED25519_SCALAR_SIZE ..< 2 * ED25519_SCALAR_SIZE)
 
         // left = kl + 8 * trunc28(zl)
-        // right = zr + kr
-
+        // right = zr + kr mod 2^256
         let left = BigUInt(Data(kl.reversed())) + BigUInt(Data(trunc256MinusGBits(zl: zl, g: g).reversed())) * BigUInt(8)
+        guard left < (BigUInt(1) << 255) else {
```
Comment (ehanoc) [inline Sources/bip32-ed25519-swift/bip32-ed25519-swift.swift:236]: Why the bitshift? Check for being less than a unsigned number composed by all 1's?  Isn't there a clearer way of doing it?

**Review by kylebeee — APPROVED** [review]
Tests all pass and everything, LGTM

**Comment by HashMapsData2Value** [comment]
@ehanoc I replaced the bitshifting with an explicit exponentiation. 2 ^ 255.

File: Sources/bip32-ed25519-swift/bip32-ed25519-swift.swift (hunk 2 above)
Comment (HashMapsData2Value) [reply Sources/bip32-ed25519-swift/bip32-ed25519-swift.swift]: Screenshot 2024-06-11 at 17 47 10

Comment (HashMapsData2Value) [reply Sources/bip32-ed25519-swift/bip32-ed25519-swift.swift]: Replaced it with the exponentiation

**Review by ehanoc — APPROVED** [review]
